### PR TITLE
Decentralized (or Distributed) Token Fund (DTF)

### PR DESCRIPTION
**Pull request by donchate**
Docs at https://github.com/donchate/steemsmartcontracts-wiki/blob/marketpools/TokenFund-Contract.md
Docs PR https://github.com/hive-engine/steemsmartcontracts-wiki/pull/10

Hive block number for activation needs to be set after reviews are complete.

### CODE REVIEW

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:63]: What is the purpose of this function?  Why do you need to replace '.' with '-' ?

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:74]: This function doesn't need to be async, as it doesn't await anything.

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:207]: If the generateDtfId function is required, then should fundId be generated through a similar process so external callers won't have to be concerned with that detail?  Same thing applies to other actions that have a fundId parameter.

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:194]: creation fee should say **update** fee

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:253]: authorperm feels to me like it's short for "author permissions".
Can you change that to something like authorpermlink, or aboutlink, to clarify the intent?

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Mostly looks good, just some minor revisions needed.  See comments.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: Updated.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: Updated.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: Fixed.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js:203]: Inherited from mining, its not required here. Removed.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: Fixed.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: authorperm naming came from hived, updated to authorPermlink to more suit our naming style

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js]: I have added a maxAccountApprovals param to scope this query to the last 1000 added approvals.

**Comment by donchate** [comment]
I spent probably more time than necessary thinking about retention of approvals and proposal weights after they are completed. \

Now, we cap at 1000 latest approvals for users, and after proposals are done the proposal.approvalweight remains static and is no longer updated. If we need to clean up more aggressively, the next step would be to remove records from the approvals table, which I think would be a substantial negative for users and frontends that wish to display a list of user approvals and past proposal support, etc. Also, I believe this would require an updated remove query in the database api to allow multiple record removal via query rather than one at a time.

File: contracts/tokenfunds.js
Comment (bt-cryptomancer) [inline contracts/tokenfunds.js:514]: I'm still a bit concerned about performance in this loop.  maxAccountApprovals is 1000, so in the worst case we might call updateProposalWeight 1000 times, and that function has a both a DB find & update, so again in the worst case that's 2000 more queries, though most likely the validation check against the proposal's endDate will cut down the updates significantly.

You could optimize it a bit by copying the endDate into the approvals table when the approval is created.  That way the information will already be available here so you can do the endDate validation without performing another DB find if the proposal has already ended.  Although if you made that change, you'd probably also want to remove the ability to update the endDate through the updateProposal action, since that would then require updating it in all the approvals table entries as well, which could have its own performance implications.  So there are trade-offs.

Also, maybe a smaller number for maxAccountApprovals would be appropriate?  1000 seems like a lot, perhaps 100 or even 50 would be better?

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Still a couple things for you to look at before I approve this, but almost done!

> If we need to clean up more aggressively, the next step would be to remove records from the approvals table, which I think would be a substantial negative for users and frontends that wish to display a list of user approvals and past proposal support, etc. Also, I believe this would require an updated remove query in the database api to allow multiple record removal via query rather than one at a time.

Fair enough, and good point about multiple record removal.  We can leave the records in the table for now and see if DB storage space ever starts becoming enough of an issue to think about methods for cleaning it up.  I think it will be a while before we really need to worry about this.

File: contracts/tokenfunds.js
Comment (donchate) [reply contracts/tokenfunds.js:514]: In my experience with the hived/DHF there have been quite a few requests for reducing the end date via an update. Proposal creators gain more flexibility if they are unsure whether they will get funded or not, and after successful funding can adjust amount and end to meet their goals without having to create a new one. It would be good to try and keep it.

An advantage of going with fewer though, such as 30 or 50, should encourage people to approve only what they believe to be higher quality proposals. It is designed to limit approvals on active proposals at any given time, so instead of having updateProposal do a large query to do the update, I am trying to update them as they go along in approveProposal and updateProposalApprovals.

Also noticed here we need to enforce the max when people are adding new approvals, to prevent any stake against an active proposal to not be updated. So I need to add more queries here, but they offer a chance to update approvals on finished proposals and verify user is not going over the max amount. If people come against this error when approving, then it will be time to consider increasing maxAccountApprovals.

So, going with 50, in either case I think it would be a max of 50 + 50 queries. (The endDate validation now doesn't save a query,  but causes the approval to be updated instead, but allows the whole thing to remain consistent).

**Review by bt-cryptomancer — APPROVED** [review]
> In my experience with the hived/DHF there have been quite a few requests for reducing the end date via an update. Proposal creators gain more flexibility if they are unsure whether they will get funded or not, and after successful funding can adjust amount and end to meet their goals without having to create a new one. It would be good to try and keep it.

Okay, this is pretty compelling reasoning, I agree it's beneficial to keep the end date updateable.

> An advantage of going with fewer though, such as 30 or 50, should encourage people to approve only what they believe to be higher quality proposals. It is designed to limit approvals on active proposals at any given time, so instead of having updateProposal do a large query to do the update, I am trying to update them as they go along in approveProposal and updateProposalApprovals.

> Also noticed here we need to enforce the max when people are adding new approvals, to prevent any stake against an active proposal to not be updated. So I need to add more queries here, but they offer a chance to update approvals on finished proposals and verify user is not going over the max amount. If people come against this error when approving, then it will be time to consider increasing maxAccountApprovals.

An approval limit of 50 feels good, and I like the approach you're taking here which should ultimately work a bit better.

I'm happy to approve this now, great work!

**Pull request merged**
